pkg/admin/migration/migration.go

```diff
@@ -0,0 +1,282 @@
+	"events",
+)
+
+var metadataAccessor = meta.NewAccessor()
+
+type migrator struct {
+	client           dynamic.Interface
+	discoveryClient  discovery.ServerResourcesInterface
+	crdClient        v1.CustomResourceDefinitionInterface
+	apiserviceClient apiregistrationv1.APIServiceInterface
+}
+
+func NewMigrator(kubeConfigPath string) (*migrator, error) {
+	restConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	crd, err := crdclient.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	apiservice, err := apiserviceclient.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	dynamic, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &migrator{
+		client:           dynamic,
+		discoveryClient:  clientset.Discovery(),
+		crdClient:        crd.ApiextensionsV1().CustomResourceDefinitions(),
+		apiserviceClient: apiservice.ApiregistrationV1().APIServices(),
+	}, nil
+}
+
+func (d *migrator) Start(ctx context.Context) (*MigrationResultList, error) {
+	return d.start(ctx)
+}
+
+func (d *migrator) start(ctx context.Context) (*MigrationResultList, error) {
+	schemas, err := d.findMigratableResources(ctx)
+	if err != nil {
+		return nil, err
+	}
+	results := &MigrationResultList{
+		Status: MigrationSuccess,
+	}
+	errorOccured := false
+	start := time.Now()
+	klog.Info("schema migration started")
+
+	wg := sync.WaitGroup{}
+	lock := sync.Mutex{}
+	wg.Add(len(schemas))
+	for _, sch := range schemas {
+		go func(wg *sync.WaitGroup, sch schema.GroupVersionResource) {
+			defer wg.Done()
+			objectList := &unstructured.UnstructuredList{}
+			var migrationErr error
+
+			migrationErr = retry.OnError(retry.DefaultBackoff, canRetry, func() error {
+				objectList, migrationErr = d.list(ctx, sch, metav1.ListOptions{})
+				if migrationErr != nil {
+					return migrationErr
+				}
+				return nil
+			})
+
+			if migrationErr != nil {
+				errorOccured = true
+				migrationErr = fmt.Errorf("could not list resources: %+v", migrationErr)
+				lock.Lock()
+				results.Items = append(results.Items, MigrationResult{Error: migrationErr, ResourceVersion: sch, Timestamp: time.Now()})
+				lock.Unlock()
+				return
+			}
+
+			for _, object := range objectList.Items {
+				ref := object
+				migrationErr := d.migrateOneItem(ctx, sch, &ref)
+				if migrationErr != nil {
+					errorOccured = true
+				}
+				lock.Lock()
+				results.Items = append(results.Items, MigrationResult{
+					Error:           migrationErr,
+					ResourceVersion: sch,
+					ObjectMeta:      getObjectMeta(&ref),
+					Timestamp:       time.Now()})
+				lock.Unlock()
+			}
+		}(&wg, sch)
+	}
+	if errorOccured {
+		results.Status = MigrationFailure
+	}
+	wg.Wait()
+	klog.Infof("schema migration finished, it took %s to complete", time.Since(start).String())
```

Comment:
BTW. Working on upgrade epic we (or I) moved to using `klog.InfoS`, trying to follow [K8s' developers logging guide](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md)

Reply:
Sounds good, updating them